local Axes = require("@self/Axes");
export type Axes = Axes.Axes;
local BrickColor = require("@self/BrickColor");
export type BrickColor = BrickColor.BrickColor;
local CFrame = require("@self/CFrame");
export type CFrame = CFrame.CFrame;
local Color3 = require("@self/Color3");
export type Color3 = Color3.Color3;
local ColorSequence = require("@self/ColorSequence");
export type ColorSequence = ColorSequence.ColorSequence;
local ColorSequenceKeypoint = require("@self/ColorSequenceKeypoint");
export type ColorSequenceKeypoint = ColorSequenceKeypoint.ColorSequenceKeypoint;
local DateTime = require("@self/DateTime");
export type DateTime = typeof(DateTime.now());
local Enum = require("@self/Enum");
export type Enum = Enum.Enum;
local EnumItem = require("@self/EnumItem");
export type EnumItem = EnumItem.EnumItem;
local Enums = require("@self/Enums");
export type Enums = Enums.Enums;
local Faces = require("@self/Faces");
export type Faces = Faces.Faces;
local Font = require("@self/Font");
export type Font = Font.Font;
local NumberRange = require("@self/NumberRange");
export type NumberRange = NumberRange.NumberRange;
local NumberSequence = require("@self/NumberSequence");
export type NumberSequence = NumberSequence.NumberSequence;
local NumberSequenceKeypoint = require("@self/NumberSequenceKeypoint");
export type NumberSequenceKeypoint = NumberSequenceKeypoint.NumberSequenceKeypoint;
local PhysicalProperties = require("@self/PhysicalProperties");
export type PhysicalProperties = PhysicalProperties.PhysicalProperties;
local Ray = require("@self/Ray");
export type Ray = Ray.Ray;
local Rect = require("@self/Rect");
export type Rect = Rect.Rect;
local UDim = require("@self/UDim");
export type UDim = UDim.UDim;
local UDim2 = require("@self/UDim2");
export type UDim2 = UDim2.UDim2;
local Vector2 = require("@self/Vector2");
export type Vector2 = Vector2.Vector2;
local Vector2int16 = require("@self/Vector2int16");
export type Vector2int16 = Vector2int16.Vector2int16;
local Vector3 = require("@self/Vector3");
export type Vector3 = Vector3.Vector3;
local Vector3int16 = require("@self/Vector3int16");
export type Vector3int16 = Vector3int16.Vector3int16;
local Content = require("@self/Content");
export type Content = Content.Content;
local UniqueId = require("@self/UniqueId");
export type UniqueId = UniqueId.UniqueId;

return {
    Axes = Axes,
    BrickColor = BrickColor,
    CFrame = CFrame,
    Color3 = Color3,
    ColorSequence = ColorSequence,
    ColorSequenceKeypoint = ColorSequenceKeypoint,
    DateTime = DateTime,
    Enum = Enum,
    EnumItem = EnumItem,
    Enums = Enums,
    Faces = Faces,
    Font = Font,
    NumberRange = NumberRange,
    NumberSequence = NumberSequence,
    NumberSequenceKeypoint = NumberSequenceKeypoint,
    PhysicalProperties = PhysicalProperties,
    Ray = Ray,
    Rect = Rect,
    UDim = UDim,
    UDim2 = UDim2,
    Vector2 = Vector2,
    Vector2int16 = Vector2int16,
    Vector3 = Vector3,
    Vector3int16 = Vector3int16,
    Content = Content,
    UniqueId = UniqueId,
};
